polkadot/node/network/collator-protocol/src/validator_side/collation.rs

```diff
@@ -267,33 +334,179 @@ impl Collations {
+	/// share depending on its assignments in the claim queue. This means that the number of
+	/// collations seconded per parachain should ideally be equal to the number of claims for the
+	/// particular parachain in the claim queue.
+	///
+	/// To achieve this each seconded collation is mapped to an entry from the claim queue. The next
+	/// fetch is the first unsatisfied entry from the claim queue for which there is an
+	/// advertisement.
+	///
+	/// If claim queue is not supported then `group_assignment` should contain just one element and
+	/// the score won't matter. In this case collations will be fetched in the order they were
+	/// received.
+	///
+	/// Note: `group_assignments` is needed just for the fall back logic. It should be removed once
+	/// claim queue runtime api is released everywhere since it will be redundant - claim queue will
+	/// already be available in `self.claim_queue_state`.
+	fn pick_a_collation_to_fetch(
+		&mut self,
+		group_assignments: &Vec<ParaId>,
+	) -> Option<(PendingCollation, CollatorId)> {
+		gum::trace!(
+			target: LOG_TARGET,
+			waiting_queue=?self.waiting_queue,
+			claims_per_para=?self.claims_per_para,
+			?group_assignments,
+			"Pick a collation to fetch."
+		);
+
+		let claim_queue_state = match self.claim_queue_state.as_mut() {
+			Some(cqs) => cqs,
+			// Fallback if claim queue is not available. There is only one assignment in
+			// `group_assignments` so fetch the first advertisement for it and return.
+			None =>
+				if let Some(assigned_para_id) = group_assignments.first() {
+					return self
+						.waiting_queue
+						.get_mut(assigned_para_id)
+						.and_then(|collations| collations.pop_front())
+				} else {
+					unreachable!("Group assignments should contain at least one element.")
+				},
+		};
+
+		let mut pending_for_para = match self.status {
+			CollationStatus::Waiting => None,
+			CollationStatus::Fetching(para_id) => Some(para_id),
+			CollationStatus::WaitingOnValidation(para_id) => Some(para_id),
+			CollationStatus::Seconded => None,
+		};
+
+		for (fulfilled, assignment) in claim_queue_state {
+			// if this assignment has been already fulfilled - move on
+			if *fulfilled {
+				continue
+			}
+
+			// if there is a pending fetch for this assignment, we should consider it satisfied and
+			// proceed with the next
+			if let Some(pending_for) = pending_for_para {
+				if pending_for == *assignment {
+					// the pending item should be used only once
+					pending_for_para = None;
+					continue
+				}
+			}
+
+			// we have found and unfulfilled assignment - try to fulfill it
+			if let Some(collations) = self.waiting_queue.get_mut(assignment) {
+				if let Some(collation) = collations.pop_front() {
+					// we don't mark the entry as fulfilled because it is considered pending
+					return Some(collation)
+				}
+			}
+		}
+
+		None
+	}
+
+	// Returns the number of pending collations for the specified `ParaId`. This function should
+	// return either 0 or 1.
```

Comment:
Should or must? If the caller is free to assume it is either those two values, it should probably be a `bool` and the name changed to `is_pending_for_para` or similar.

Reply:
Done